Step3 - 로또(자동)

Kelicia91 opened the pull request.

Kelicia91 committed:
docs: 기능 목록 추가

Kelicia91 committed:
chore: delete empty.txt

Kelicia91 committed:
docs: 기능 목록 보강

Kelicia91 committed:
feat: LottoNumber

Kelicia91 committed:
feat: Lotto

Kelicia91 committed:
feat: LottoStore

Kelicia91 committed:
refactor: rename method

Kelicia91 committed:
docs: 기능 목록에 lotto parser 추가

Kelicia91 committed:
feat: LottoParser

Kelicia91 committed:
fix: missing Lotto validation (numbers size, distinct)

Kelicia91 committed:
feat: WinningLotto

Kelicia91 committed:
feat: LottoWinningCriteria

Kelicia91 committed:
feat: LottoWinningStatistics

Kelicia91 committed:
refactor: extends exceptions

Kelicia91 committed:
fix: duplicated lottos

Kelicia91 committed:
refactor: gardening

Kelicia91 committed:
refactor: rename Criteria to LottoWinningRank

Kelicia91 committed:
feat: InputView, OutputView

Kelicia91 committed:
feat: LottoApp

Kelicia91 committed:
refactor: rename param

minsungryu commented on src/main/java/nextstep/utils/Console.java:
```diff
@@ -0,0 +1,63 @@
+ * @version 1.0
+ * @since 1.0
+ */
+public class Console {
```
오잉 이건 직접 작성하신 코드는 아닌걸까요?!

Kelicia91 replied:
아 프리코스 때 사용자 입력은 nextstep.utils 패키지에서 제공하는 api를 활용하라는 제약사항이 있었어서 그대로 가져왔어요~

minsungryu commented on src/main/java/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+	private static List<LottoNumber> generate() {
+		Collections.shuffle(CANDIDATE_LOTTO_NUMBERS);
+		final List<LottoNumber> numbers = new ArrayList<>(CANDIDATE_LOTTO_NUMBERS.subList(0, NUM_OF_LOTTO_NUMBERS));
+		numbers.sort(Comparator.comparingInt(LottoNumber::get));
```
`LottoNumber`가 `Comparable` 인터페이스를 구현해볼 수도 있을 것 같아요

Kelicia91 replied:
오 츄라이해볼게요!

minsungryu commented on src/main/java/LottoApp.java:
```diff
@@ -0,0 +1,60 @@
+		return lotto;
+	}
+
+	private List<Lotto> buyLottoAt(LottoStore store) {
+		try {
+			return store.sell(inputView.payKRW());
```
1) `store`는 `LottoApp`의 멤버 변수인데 인자로 넘겨 받으시는 이유가 있을까요?
2) `inputView.payKRW()`을 직접 호출하는 것보다는 인자로 `int payKRW`를 받아보면 어떨까 싶습니다. `run()` 메서드에서 일관되게 input => 행위 => output 동작을 기대할 수 있을 것 같아요.

Kelicia91 replied:
1. 엄.. 딱히 별 이유는 없는데; 주입하는게 뭔가 자연스러워서(??) 저렇게 해둔것 같네요. 굳이 불필요하게 주입안하고 멤버변수를 그대로 활용해도 될 것 같네요. 수정해둘게요!
2. 아 그게 더 좋겟네요 👍 수정해둘게요!

minsungryu commented on src/main/java/InputView.java:
```diff
@@ -0,0 +1,20 @@
+	private final OutputView outputView;
+
+	public InputView(OutputView outputView) {
+		this.outputView = outputView;
```
입력을 받기 위해 메시지를 화면에 출력한다는 행위면에서는 이해가 되지만, 객체지향적인 관점에서는 `InputView`가 `OutputView`에 의존해야할까? 고민이 됩니다.

Kelicia91 replied:
Aㅏ.. 저도 이거 매우 의문스러운 부분이었는데요, 리뷰 코멘트랑 코드 다시 읽어보니 '`InputView`가 정말 필요한가?' 싶은 생각이 드네요. `LottoApp` 리팩토링하면서 정리해볼게요

minsungryu commented on src/main/java/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+		}
+	}
+
+	public int getNumOfLottoNumbers() {
```
테스트 관점에서 작성된 메서드라는 생각이 들어요.
로또가 6개의 숫자를 가져야한다면, 객체 내부 값이 6개가 들어있는지 확인하는 것보다 생성자를 통해서 잘못된 객체가 생성이 안됨을 검증해보는 것은 어떨까요?

Kelicia91 replied:
오우.. 코드 다시보니 '....왜 이런걸 만들엇지???' 라는 생각이 드네요 ㅋㅋ; ㅠㅜ
getter 지우고 검증을 보강해볼게요!

minsungryu commented on src/main/java/LottoWinningStatistics.java:
```diff
@@ -0,0 +1,45 @@
+	}
+
+	public float earningRate() {
+		final int prizeKRW = numOfLottosByMatching.entrySet().stream()
```
sum을 하게 되면 int 범위를 초과할 위험이 있어보입니다. long으로 선언하시는 건 어떨까요?

Kelicia91 replied:
아하 수정해둘게요!

minsungryu commented on src/main/java/WinningLotto.java:
```diff
@@ -0,0 +1,20 @@
+	public int countMatching(Lotto lotto) {
+		int count = 0;
+		for (LottoNumber lottoNumber : lotto.lottoNumbers) {
+			count += (this.lottoNumbers.contains(lottoNumber) ? 1 : 0);
```
for-each 문에서 부수효과를 일으키는 대신 Stream API로 변경해보는 것은 어떨까요??

Kelicia91 replied:
오 츄라이해보겟슴니다

minsungryu commented on src/main/java/LottoWinningStatistics.java:
```diff
@@ -0,0 +1,45 @@
+				, (acc, entry) -> acc + calculatePrizeKRW(entry.getKey(), entry.getValue())
+				, Integer::sum);
+
+		final int paidKRW = Lotto.PRICE_KRW * (numOfLottosByMatching.values().stream()
```
괄호 안에 있는 스트림 식은 변수로 빼면 의도가 좀 더 명확히 보여서 가독성이 좋아질 것 같습니다!

Kelicia91 replied:
아 그게 더 좋겟네요. 네 수정해둘게요!

minsungryu reviewed (changes requested):
지우님 로또 미션 잘 구현해주셨네요 👍 
제가 몇 가지 코멘트를 남겼는데 확인해주시면 좋을 것 같습니다.

minsungryu commented:
화이팅입니다

Kelicia91 committed:
refactor: LottoNumber implements Comparable

Kelicia91 committed:
refactor: LottoApp, View

- 메소드에 LottoStore 주입을 멤버변수로 대체
- view, logic 분리
- 불필요한 InputView 삭제하고 View로 대체

Kelicia91 committed:
refactor: rename methods - View의 in, out 분리

Kelicia91 committed:
refactor: 테스트 코드에서만 활용되는 Lotto.getNumOfLottoNumber() 제거

- test: 로또 번호 갯수 검증 테스트에서 6개 보다 큰 경우 추가

Kelicia91 committed:
refactor: 총상금 data type 변경(int->long)

- refactor: LottoWinningStatisticsTest 메소드별로 분리
- refactor: LottoWinningStatisticsTest 당첨 랭크별로 테스트 데이터를 생성해 테스트 케이스 추가

Kelicia91 committed:
refactor: WinningLotto.countMatching() using stream

Kelicia91 committed:
refactor: '로또 갯수 구하기' 메소드 분리

Kelicia91 commented:
코드 양이 적지 않음에도 불구하고 꼼꼼히 리뷰해주셔서 감사합니다 🙇‍♀️ 
코멘트 주셨던 부분들 반영했고 확인 부탁드릴게요~

Kelicia91 committed:
refactor: 불필요한 Comparator 제거

- test: LottoParserTest - parse 결과 요소 검증, 입력 길이가 긴 경우 검증 추가

minsungryu reviewed (approved):
지우님, 코드도 너무 깔끔하고 멋지네요 LGTM 💯 
3단계 구현하시느라 고생 많으셨습니다.
바로 merge 할테니 다음 미션도 잘 부탁드려요.
